### ISSUE TITLE
[git] Frontend tries to watch the status of a non-existing repository after startup

### ISSUE THREAD

**Issue by kittaakos** [theia-ide/theia#638]
- Start application.
 - Clone a git repository into your workspace.
 - Refresh the UI so that client will be informed of the new repository.
 - Kill application.
 - Delete the cloned repository outside from Theia.
 - Restart the application.

It tries to watch the non-existing repository which throws an error each second. Even if the client state restoration is not super sophisticated, the watcher should be smart enough to unregister itself if the target repository does not exist anymore.

**Comment by kittaakos** [comment]
Same when you just rename the Git repository directory to something else.


For the initial issue, the following changes will be sufficient:
```diff
diff --git a/packages/git/src/node/dugite-git-watcher.ts b/packages/git/src/node/dugite-git-watcher.ts
index 501396d2..236f76e4 100644
--- a/packages/git/src/node/dugite-git-watcher.ts
+++ b/packages/git/src/node/dugite-git-watcher.ts
@@ -33,16 +33,22 @@ export class DugiteGitWatcherServer implements GitWatcherServer {
         const watcher = this.watcherSequence++;
         const interval = this.preferences['git.pollInterval']; // TODO refresh timers on preference change.
         const timer = setInterval(async () => {
-            const status = await this.git.status(repository);
-            const oldStatus = this.status.get(repository);
-            if (this.client && !WorkingDirectoryStatus.equals(status, oldStatus)) {
-                this.status.set(repository, status);
-                const event: GitStatusChangeEvent = {
-                    source: repository,
-                    status,
-                    oldStatus
-                };
-                this.client.onGitChanged(event);
+            try {
+                const status = await this.git.status(repository);
+                const oldStatus = this.status.get(repository);
+                if (this.client && !WorkingDirectoryStatus.equals(status, oldStatus)) {
+                    this.status.set(repository, status);
+                    const event: GitStatusChangeEvent = {
+                        source: repository,
+                        status,
+                        oldStatus
+                    };
+                    this.client.onGitChanged(event);
+                }
+            } catch (error) {
+                if (error.message === 'Unable to find path to repository on disk.') {
+                    await this.unwatchGitChanges(watcher);
+                }
             }
         }, interval);
         this.watchers.set(watcher, timer);
@@ -50,9 +56,12 @@ export class DugiteGitWatcherServer implements GitWatcherServer {
     }

     async unwatchGitChanges(watcher: number): Promise<void> {
-        if (!this.watchers.delete(watcher)) {
+        const timer = this.watchers.get(watcher);
+        if (!timer) {
             throw new Error(`No Git watchers were registered with ID: ${watcher}.`);
         }
+        clearInterval(timer);
+        this.watchers.delete(watcher);
     }

     dispose(): void {
@@ -63,4 +72,4 @@ export class DugiteGitWatcherServer implements GitWatcherServer {
         this.client = client;
     }

-}
\ No newline at end of file
+}
```

**Comment by kittaakos** [comment]
This needs adjustments on the frontend too, git-repository-provider tries to stick to the previous name although the backend provides the new name.

That results in an exception when trying to fetch even if I have refreshed the git widget.
```
messages.js:46 Uncaught (in promise) Error: Request add failed with message: Unable to find path to repository on disk.
    at new ResponseError (messages.js:46)
    at handleResponse (main.js:421)
    at processMessageQueue (main.js:249)
    at main.js:233
    at run (setImmediate.js:40)
    at runIfPresent (setImmediate.js:69)
    at onGlobalMessage (setImmediate.js:109)
```

**Comment by kittaakos** [comment]
```diff
diff --git a/packages/git/src/browser/git-repository-provider.ts b/packages/git/src/browser/git-repository-provider.ts
index 3733a14d..c310056b 100644
--- a/packages/git/src/browser/git-repository-provider.ts
+++ b/packages/git/src/browser/git-repository-provider.ts
@@ -11,7 +11,7 @@ import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service
 @injectable()
 export class GitRepositoryProvider {

-    protected selectedRepository: Repository;
+    protected selectedRepository: Repository | undefined;

     constructor(
         @inject(Git) protected readonly git: Git,
@@ -27,7 +27,7 @@ export class GitRepositoryProvider {
         }
     }

-    select(localUri: string) {
-        this.selectedRepository = { localUri };
+    select(localUri: string | undefined) {
+        this.selectedRepository = localUri ? { localUri } : undefined;
     }
 }
diff --git a/packages/git/src/browser/git-widget.ts b/packages/git/src/browser/git-widget.ts
index 005ceb87..f46152a5 100644
--- a/packages/git/src/browser/git-widget.ts
+++ b/packages/git/src/browser/git-widget.ts
@@ -152,7 +152,13 @@ export class GitWidget extends VirtualWidget {
         const refresh = h.a({
             className: 'button',
             title: 'Refresh',
-            onclick: e => {
+            onclick: async e => {
+                const selected = await this.gitRepositoryProvider.getSelected();
+                try {
+                    await this.git.status(selected);
+                } catch {
+                    this.gitRepositoryProvider.select(undefined);
+                }
                 this.initialize();
             }
         }, h.i({ className: 'fa fa-refresh' }));
```

**Comment by hexa00** [comment]
Note this is also the case if you just start with a workspace that is not a git repo and click on the Git tab.

**Comment by kittaakos** [comment]
Merged into #621.

**Comment by kittaakos** [comment]
> Note this is also the case if you just start with a workspace that is not a git repo and click on the Git tab.

Will be available with 9c6f50ae8aa828622962d08168a3d0e3bfae445b.
